tun/tun_illumos.go

```diff
@@ -0,0 +1,271 @@
+		return 0, err
+	default:
+		_, read, _, err := unix.Getmsg(int(tun.tunFile.Fd()), nil, buf[offset:])
+		if err != nil {
+			return 0, err
+		}
+		return len(read), nil
+	}
+}
+
+// Write implements the Device interface
+func (tun *NativeTun) Write(buf []byte, offset int) (int, error) {
+	if err := unix.Putmsg(int(tun.tunFile.Fd()), nil, buf[offset:], 0); err != nil {
+		return 0, err
+	}
+	return len(buf), nil
+}
+
+// File implements the Device interface
+func (tun *NativeTun) File() *os.File {
+	return tun.tunFile
+}
+
+// MTU implements the Device interface
+func (tun *NativeTun) MTU() (int, error) {
+	return tun.mtu, nil
+}
+
+// Events implements the Device interface
+func (tun *NativeTun) Events() chan Event {
+	return tun.events
+}
+
+// Close implements the Device interface
+func (tun *NativeTun) Close() error {
+	if tun.ip_fd >= 0 {
+
+		var l unix.Lifreq
+		if err := l.SetName(tun.name); err != nil {
+			return err
+		}
+
+		// illumos defines the ioctl number as a signed int, but the
+		// common x/sys/unix functions use uint. Hand-cast things to help
+		// the compiler figure it out.
+		reqnum := int(unix.SIOCGLIFMUXID)
+		if err := unix.IoctlLifreq(tun.ip_fd, uint(reqnum), &l); err != nil {
+			return err
+		}
+
+		id := l.GetLifruInt()
+
+		if err := unix.IoctlSetInt(tun.ip_fd, unix.I_PUNLINK, id); err != nil {
+			return err
+		}
+
+		unix.Close(tun.ip_fd)
+		tun.ip_fd = -1
+	}
+
+	if tun.tunFile != nil {
+		tun.tunFile.Close()
+		tun.tunFile = nil
+	}
+
+	if tun.events != nil {
+		close(tun.events)
+	}
+
+	return nil
+}
+
+func CreateTUNFromFile(file *os.File, mtu int) (Device, error) {
+	/*
+	 * XXX It's not currently clear to me how to take a TUN file descriptor
+	 * and determine the attached PPA.
+	 *
+	 * XXX Based on the current architecture of the daemon, we need to be
+	 * able to reconstitute our NativeTun object from just a file
+	 * descriptor number as the fd number is passed (with no other
+	 * information) through the environment to the child.  For now, use the
+	 * "-f" flag.
+	 */
+	return nil, fmt.Errorf("CreateTUNFromFile() not currently supported")
+}
+
+func CreateTUN(name string, mtu int) (Device, error) {
+	if name != "tun" {
+		return nil, fmt.Errorf("Interface name must be 'tun'")
+	}
+
+	/*
+	 * To establish a "tun" interface, we need to open a few file
+	 * descriptors.
+	 */
+	ip_node := "/dev/udp"
+	dev_node := "/dev/tun"
+
+	/*
+	 * First, the IP driver:
+	 */
+	ip_fd, err := unix.Open(ip_node, unix.O_RDWR, 0)
+	if err != nil {
+		return nil, fmt.Errorf("Could not open IP (%s)", ip_node)
+	}
+
+	/*
+	 * Now, the TUN driver.  Note that we use "os.OpenFile()" instead of
+	 * "unix.OpenFile()" so that we get the os.File object here.  The rest
+	 * of the Wireguard API seems to depend on that functionality.
+	 */
+	tunFile, err := os.OpenFile(dev_node, unix.O_RDWR, 0)
+	if err != nil {
+		unix.Close(ip_fd)
+		return nil, fmt.Errorf("Could not open TUN (%s)", dev_node)
+	}
+	fd := int(tunFile.Fd())
```

Comment:
Cache the SyscallConn() and use that, and avoid Fd(), perhaps, as discussed above.